### PR TITLE
Experiment with read retries to handle BufferUnderflowExceptions.

### PR DESCRIPTION
**Pull request by metasim**
To address #131

### CODE REVIEW

**Comment by courtney-whalen** [comment]
[summer_crop_model (2).json.zip](https://github.com/locationtech/rasterframes/files/3249069/summer_crop_model.2.json.zip)

**Comment by metasim** [comment]
According to @courtney-whalen large model training completed after this change.

**Comment by courtney-whalen** [comment]
Yeah, tested it out several times and haven't gotten a BufferUnderflowException again

File: core/src/main/scala/org/locationtech/rasterframes/ref/RasterSource.scala [hunk 1]
```diff
@@ -221,21 +239,49 @@ object RasterSource extends LazyLogging {
...
+//              throw be
+//            }
+//            else Thread.sleep(100)
+//            ()
+//        }
+//      }
+//      result
+//    }
+
+
```
Comment (metasim) [inline core/src/main/scala/org/locationtech/rasterframes/ref/RasterSource.scala]: Sorry about that comment block. Will get rid of eventually.

File: core/src/main/scala/org/locationtech/rasterframes/ref/RasterSource.scala [hunk 2]
```diff
@@ -191,7 +191,25 @@ object RasterSource extends LazyLogging {
...
+      catch {
+        // On this exeception we attempt to recreate the delegate and read again.
+        case be: java.nio.BufferUnderflowException =>
+          _delRef = null
+          val newDel = delegateBuilder()
+          val result = f(newDel)
+          _delRef = newDel
+          result
+      }
+    }
```
Comment (metasim) [inline core/src/main/scala/org/locationtech/rasterframes/ref/RasterSource.scala]: This is the tricky bit. Need to make sure it's thread safe, even though multiple threads "shouldn't" be calling the read code.

**Comment by pomadchin** [comment]
@metasim did it finally work?

**Comment by metasim** [comment]
@pomadchin Not sure yet.... takes a larger deployment process and several hours of training to be absolutely sure.

**Pull request merged**